spirv_reflect.cpp

```diff
@@ -0,0 +1,514 @@
+		SPIRV_CROSS_THROW("Invalid JSON state");
+
+	if (json_state.top().second)
+		statement_inner(",\n");
+	statement_no_return("\"", key, "\" : ");
+	json_state.top().second = true;
+}
+
+void CompilerReflection::emit_json_key_value(const std::string &key, const std::string &value)
+{
+	emit_json_key(key);
+	statement_inner("\"", value, "\"");
+}
+
+void CompilerReflection::emit_json_key_value(const std::string &key, uint32_t value)
+{
+	emit_json_key(key);
+	statement_inner(value);
+}
+
+void CompilerReflection::emit_json_key_value(const std::string &key, bool value)
+{
+	emit_json_key(key);
+	statement_inner(value ? "true" : "false");
+}
+
+void CompilerReflection::emit_json_key_object(const std::string &key)
+{
+	emit_json_key(key);
+	statement_inner("{\n");
+	++indent;
+	json_state.emplace(JsonType::Object, false);
+}
+
+void CompilerReflection::emit_json_key_array(const std::string &key)
+{
+	emit_json_key(key);
+	statement_inner("[\n");
+	++indent;
+	json_state.emplace(JsonType::Array, false);
+}
+
+string CompilerReflection::compile()
+{
+	// force a GLSL compilation in the parent class in order to set all our internal state
+	CompilerGLSL::compile();
+
+    buffer.reset();
+
+	// Force a classic "C" locale, reverts when function returns
+	ClassicLocale classic_locale;
+
+	// Move constructor for this type is broken on GCC 4.9 ...
+	buffer = unique_ptr<ostringstream>(new ostringstream());
+	begin_json_object();
+	emit_extensions();
+	emit_types();
+	emit_resources();
+	end_json_object();
+
+	return buffer->str();
+}
+
+void CompilerReflection::emit_extensions()
+{
+	if (forced_extensions.empty())
+		return;
+
+	emit_json_key_array("extensions");
+	for (const auto &ext : forced_extensions)
+		emit_json_array_value(ext);
+	end_json_array();
+}
+
+void CompilerReflection::emit_types()
+{
+	bool emitted_open_tag = false;
+	for (auto &id : ids)
+	{
+		if (id.get_type() == TypeType)
+		{
+			auto &type = id.get<SPIRType>();
+			if (type.basetype == SPIRType::Struct && type.array.empty() && !type.pointer)
+			{
+				emit_type(type, emitted_open_tag);
+			}
+		}
+	}
+
+	if (emitted_open_tag)
+	{
+		end_json_object();
+	}
+}
+
+void CompilerReflection::emit_type(const SPIRType &type, bool &emitted_open_tag)
+{
+	auto name = type_to_glsl(type);
+
+	// Struct types can be stamped out multiple times
+	// with just different offsets, matrix layouts, etc ...
+	// Type-punning with these types is legal, which complicates things
+	// when we are storing struct and array types in an SSBO for example.
+	// If the type master is packed however, we can no longer assume that the struct declaration will be redundant.
+	if (type.type_alias != 0 && !has_decoration(type.type_alias, DecorationCPacked))
+		return;
+
+	//add_resource_name(type.self);
+	//type.member_name_cache.clear();
+
+	if (!emitted_open_tag)
+	{
+		emit_json_key_object("types");
+		emitted_open_tag = true;
+	}
+	emit_json_key_array(name);
```

Comment:
Emitting the types based on name is dangerous. SPIR-V modules can have identical names for multiple unrelated types. It's very common to have two types with same name, one living as part of an SSBO, one as a value type. While we need to deal with this in GLSL/HLSL/MSL/etc, this output should just be a "raw" reflection of the SPIR-V module.

I think it makes more sense to have "types" as a flat array, and then objects will refer to types by index into that array (or some other way to index the objects). The name would be part of the inner object in the array. This way there is no ambiguity.

Consider the problem:
```
"types" : [
    "foo" : { "do something" },
    "foo" : { "do something else" }
]
```

Reply:
using indexing would require maintaining additional internal state between different sections of the JSON emission code.  Would it be reasonable to use the type id (rather than the name) as the key into a map of types?  Also, again, are there any examples in the test data of this kind of name duplication, so I can validate my fix?

Reply:
Type ID lookup is also fine in my book. See shaders/comp/struct-packing.comp for some difficult cases.